File: 2025/src/day_01.py
from pathlib import Path
from typing import NamedTuple

from utils import benchmark


class Data(NamedTuple):
    lines: list[str]


@benchmark
def parse() -> Data:
    with open(Path(__file__).parent.parent / "inputs/day_01.txt") as f:
        file = f.read().strip().splitlines()
    return Data(file)


@benchmark
def part_1(data: Data) -> None:
    result = 0
    curr_pos = 50

    for line in data.lines:
        sign = 1 if line[0] == "R" else -1
        num = int(line[1:])

        curr_pos = (curr_pos + (sign * num)) % 100
        if curr_pos == 0:
            result += 1

    print(f"Day 01, Part 1: {result}")


@benchmark
def part_2(data: Data) -> None:
    result = 0
    curr_pos = 50

    for line in data.lines:
        sign = 1 if line[0] == "R" else -1
        num = int(line[1:])

        for _ in range(num):
            curr_pos = (curr_pos + sign) % 100
            if curr_pos == 0:
                result += 1

    print(f"Day 01, Part 2: {result}")


if __name__ == "__main__":
    data = parse()
    part_1(data)
    part_2(data)

File: 2025/src/day_01.rs
use std::{char, fs};

fn part_1(lines: &Vec<(char, i16)>) {
    let mut result = 0;
    let mut dial = 50;

    for (dir, num) in lines {
        dial = (dial + (if *dir == 'L' { -1 } else { 1 } * num)) % 100;
        if dial == 0 {
            result += 1;
        }
    }

    println!("Day 01, Part 1: {}", result);
}

fn part_2(lines: &Vec<(char, i16)>) {
    let mut result = 0;
    let mut dial = 50;

    for (dir, num) in lines {
        for _ in 0..*num {
            dial = (dial + (if *dir == 'L' { -1 } else { 1 })) % 100;
            if dial == 0 {
                result += 1;
            }
        }
    }
    println!("Day 01, Part 2: {}", result);
}

fn main() {
    let file = fs::read_to_string("inputs/day_01.txt").expect("error reading file");
    let lines: Vec<(char, i16)> = file
        .lines()
        .map(|l| {
            let num = l[1..].parse::<i16>().unwrap();
            return (l.chars().next().unwrap(), num);
        })
        .collect();

    part_1(&lines);
    part_2(&lines);
}
